includes/Modules/Idea_Hub.php

```diff
@@ -730,4 +728,89 @@ private function is_idea_post( $post_id ) {
 		return is_array( $this->get_post_idea( $post_id ) );
 	}
 
+	/**
+	 * Gets the parent slug to use for Idea Hub API requests.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string Parent slug.
+	 */
+	private function get_parent_slug() {
+		$reference_url = $this->context->get_reference_site_url();
+		$reference_url = rawurlencode( $reference_url );
+
+		$parent = "platforms/sitekit/properties/{$reference_url}";
+
+		return $parent;
+	}
+
+	/**
+	 * Fetches ideas from the Idea Hub API.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $type Ideas type. Valid values "saved", "new" or an empty string which means all ideas.
+	 * @return mixed List ideas request.
+	 */
+	private function fetch_ideas( $type ) {
+		$parent = $this->get_parent_slug();
+		$params = array(
+			'pageSize' => 100,
+		);
+
+		if ( 'saved' === $type ) {
+			$params['filter'] = 'saved(true)';
+		} elseif ( 'new' === $type ) {
+			$params['filter'] = 'saved(false)';
+		}
+
+		$service = $this->get_service( 'ideahub' );
+		$ideas   = $service->platforms_properties_ideas;
+
+		return $ideas->listPlatformsPropertiesIdeas( $parent, $params );
+	}
+
+	/**
+	 * Filters out ideas for which we have already created a post.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param array $ideas Ideas list to filter.
+	 * @return array Filtered ideas list.
+	 */
+	private function filter_out_drafted_ideas( $ideas ) {
+		global $wpdb;
+
+		// Return early if there are no ideas in the incoming array.
+		if ( empty( $ideas ) ) {
+			return $ideas;
+		}
+
+		$names = wp_list_pluck( $ideas, 'name' );
+
+		// phpcs:ignore WordPress.DB.DirectDatabaseQuery
+		$drafted_ideas = $wpdb->get_col(
+			// phpcs:ignore WordPress.DB.PreparedSQLPlaceholders.ReplacementsWrongNumber
+			$wpdb->prepare(
+				"SELECT DISTINCT meta_value FROM {$wpdb->postmeta} WHERE meta_key = %s AND meta_value IN (" . implode( ', ', array_fill( 0, count( $names ), '%s' ) ) . ')',
+				Post_Idea_Name::META_KEY,
+				...$names
+			)
+		);
```

Comment:
While this is clever (and I don't have any problem bypassing object cache here), I think we should use a similar approach here to what we're doing for `GET:draft-post-ideas` (and the similar `GET:published-post-ideas`).

For example, what you have here now ignores whether the post is a draft, it will get this data for _any_ post that has it - including e.g. trashed posts. We should ensure that we use a single approach for getting ideas that have a post throughout to avoid any inconsistencies.

I'd be in favor of just using the established `WP_Query` approach using a `meta_query` rather than putting together a direct database query - I don't think the performance benefit of the simplified query is worth it, and on the other end it would mean that object caching isn't used when available. How about we introduce a private method `get_idea_posts( $status )` which we then use in here, in `GET:draft-post-ideas` and in `GET:published-post-ideas`?

Reply:
> I think we should use a similar approach here to what we're doing for GET:draft-post-ideas (and the similar GET:published-post-ideas).
> We should ensure that we use a single approach for **getting ideas that have a post** throughout to avoid any inconsistencies.
> How about we introduce a private method get_idea_posts( $status ) which we then use in here, in GET:draft-post-ideas and in GET:published-post-ideas?

The problem is that `GET:draft-post-ideas` and `GET:published-post-ideas` datapoints **get posts that have an idea**, not the other way around. Therefore we can't introduce a private method that will use one algorithm for all `{draft|published}-post-ideas` and `{new|saved}-ideas` datapoints.

> I'd be in favor of just using the established WP_Query approach using a meta_query rather than putting together a direct database query - I don't think the performance benefit of the simplified query is worth it, and on the other end it would mean that object caching isn't used when available.

Currently, this method adds one SQL query per request to `new-ideas`/`saved-ideas` endpoints. If we rework it to use `WP_Query`, it will add `1 + N` SQL queries for the first request (where `N` is the number of ideas, the current limit is `100`) when object cache is not used (usually on shared hostings) and the same `1` query for sequential requests (because queries created by WP_Query are not cached) when object cache is used on the server. Plus, the query created by WP_Query will be always much slower than the direct query to the postmeta table.

So, I think it would be better if we use the direct query to the postmeta table as it is done right now. But if you still want me to rework it to use the WP_Query approach, let me know and I'll update it.

Reply:
> For example, what you have here now ignores whether the post is a draft, it will get this data for any post that has it - including e.g. trashed posts.

Forgot to comment on this. If I created a post and then decided to move it to trash, it means that I am not interested in that idea anymore, right? I think we don't need to care about post statuses when we filter new and saved ideas.

Reply:
@eugene-manuilov 
> > How about we introduce a private method get_idea_posts( $status ) which we then use in here, in GET:draft-post-ideas and in GET:published-post-ideas?
> 
> The problem is that `GET:draft-post-ideas` and `GET:published-post-ideas` datapoints **get posts that have an idea**, not the other way around. Therefore we can't introduce a private method that will use one algorithm for all `{draft|published}-post-ideas` and `{new|saved}-ideas` datapoints.

I don't see why that is a limitation. For `GET:new-ideas` and `GET:saved-ideas`, we could query the ("draft" or "publish") posts that have an idea, loop through them and remove them from the ideas list.

> > I'd be in favor of just using the established WP_Query approach using a meta_query rather than putting together a direct database query - I don't think the performance benefit of the simplified query is worth it, and on the other end it would mean that object caching isn't used when available.
> 
> Currently, this method adds one SQL query per request to `new-ideas`/`saved-ideas` endpoints. If we rework it to use `WP_Query`, it will add `1 + N` SQL queries for the first request (where `N` is the number of ideas, the current limit is `100`) when object cache is not used (usually on shared hostings) and the same `1` query for sequential requests (because queries created by WP_Query are not cached) when object cache is used on the server. Plus, the query created by WP_Query will be always much slower than the direct query to the postmeta table.

I'm not sure the "1 + N" is entirely accurate. When `WP_Query` queries 100 posts, it will typically make 2 queries, with or without persistent object cache:
1. Get the IDs for all posts that satisfy the query.
2. Query all data for those ideas to prime the relevant post caches. Without object cache, this is still cached in a variable, so no subsequent queries are made, i.e. that's where I think the "1 + N" is not accurate.

> Forgot to comment on this. If I created a post and then decided to move it to trash, it means that I am not interested in that idea anymore, right? I think we don't need to care about post statuses when we filter new and saved ideas.

We shouldn't use that as a signal that the idea isn't relevant anymore. Maybe the user wrote a post, thought it was bad and trashed it to write another one later. Specifically for both `GET:new-ideas` and `GET:saved-ideas`, we should filter out any ideas for which there is a post, no matter whether it's draft or publish (we also need to ensure a published post's idea doesn't still appear in "New" or "Saved"). Technically, we should only exclude trashed posts, since those cannot be edited and it should be treated like a deleted post.

I think we should go with the `WP_Query` approach, so that we
* Can take the post status into account, which we should definitely do
* Keep all queries around posts with an idea follow the same approach to avoid inconsistent data
* Follow the WP best practice to not use direct DB queries unless absolutely necessary (this is the least important point, but I don't think the performance benefit here is relevant enough, for the above reason)

I think the most flexible approach would be a private method like `query_idea_posts( array $args )` where you can pass extra arguments for querying that will be merged into defaults. The defaults could be what's currently provided in `GET:published-post-ideas`. We can then use that method as follows:
* In `GET:new-ideas` and `GET:saved-ideas` query posts of any post status other than "trash".
* In `GET:draft-post-ideas` query posts of the "draft" post status.
* In `GET:published-post-ideas` query posts of the "publish" post status.

Reply:
@felixarntz I think you are right, thanks for the detailed answer. Updated. :+1: